flex_model/core/wrapper.py

```diff
@@ -59,13 +61,170 @@ class _SharedState:
     offload_mode: str
 
 
+class _HookFunctionGroupManager:
+    def __init__(self):
+        self.hook_fn_to_groups_map = defaultdict(set)
+        self.groups = set()
+
+    def get_hook_fn_groups(self, hook_fn: HookFunction) -> Set[str]:
+        return self.hook_fn_to_groups_map[hook_fn]
+
+    def get_group_hook_fns(self, group_name: str) -> List[HookFunction]:
+        hook_fns = []
+        for hook_fn, groups in self.hook_fn_to_groups_map.items():
+            if group_name in groups:
+                hook_fns.append(hook_fn)
+        return hook_fns
+
+    def create(
+        self,
+        group_name: str,
+        group_constructor: str,
+        expected_shape: Tuple[Optional[int], ...],
+        all_names: List[str],
+        editing_function: Optional[Callable] = None,
+        unpack_idx: int = 0,
+    ) -> List[HookFunction]:
+        """Instantiate and group hook functions based on module name pattern matching."""
+        matching_modules = []
+        for module_name in all_names:
+            if group_constructor in module_name:
+                matching_modules.append(module_name)
+
+        hook_fns = []
+        for mod_name in matching_modules:
+            hf = HookFunction(
+                module_name=mod_name,
+                expected_shape=expected_shape,
+                editing_function=editing_function,
+                unpack_idx=unpack_idx,
+            )
+            self.hook_fn_to_groups_map[hf].add(group_name)
+            hook_fns.append(hf)
+
+        if group_name not in self.groups:
+            self.groups.add(group_name)
+
+        return hook_fns
+
+    @functools.singledispatchmethod
+    def update(self, group_constructor, group_name) -> None:
+        raise NotImplementedError(f"Cannot make group using: {type(group_constructor)}")
+
+    @update.register
+    def _update_by_list_hook_fns(
+        self, group_constructor: list, group_name: str
+    ) -> None:
+        """Add group tag to a collection of hook functions."""
+        are_hook_fns = map(lambda x: isinstance(x, HookFunction), group_constructor)
+        assert all(are_hook_fns), f"set_group takes a collection of only HookFunctions"
+
+        # Associate the hook functions with this group.
+        for hook_fn in group_constructor:
+            self.hook_fn_to_groups_map[hook_fn].add(group_name)
+
+        if group_name not in self.groups:
+            self.groups.add(group_name)
+
+    @update.register
+    def _update_by_hook_fn(
+        self, group_constructor: HookFunction, group_name: str
+    ) -> None:
+        """Add a group tag to a single hook function."""
```

Comment:
Stick to convention of either having or not having docstrings for private methods and functions.

Reply:
For private classes/methods having optional docstrings should be fine